toster3d/YMP-deluxe-edition
# FastAPI refactoring

# Migration from Flask to FastAPI with PostgreSQL Integration

Closes #6

## Overview
This PR introduces a major architectural upgrade by transitioning from Flask to FastAPI and migrating from SQLite to PostgreSQL, aiming to improve performance and scalability.

## Key Changes
- Replaced Flask with FastAPI for async request handling
- Migrated database from SQLite to PostgreSQL 
- Implemented async Redis integration for token management
- Updated Docker Compose configuration for new architecture
- Introduced Pydantic for type-safe configuration and data validation

## Detailed Changes

### Docker Compose Adjustments
- Removed Flask-related environment variables
- Updated database configuration for PostgreSQL
- Added health checks for PostgreSQL
- Introduced volumes for PostgreSQL data persistence

### Application Entry Point
- Replaced Flask with FastAPI framework
- Updated imports and configuration settings
- Integrated async database initialization 
- Added CORS middleware
- Refactored application setup for FastAPI structure

### Configuration
- Implemented Pydantic BaseSettings for type-safe configuration
- Added structured settings for database, JWT and Redis
- Streamlined configuration process
- Enhanced environment variables handling

### Dependencies Module
- Introduced async Redis integration
- Implemented robust error handling
- Added RedisTokenStorage dependency
- Configured Redis client parameters

### Extensions Module
- Implemented async SQLAlchemy capabilities
- Configured async database engine
- Added async session management
- Implemented async context managers

### JWT Utilities
- Implemented JWT token creation/verification
- Integrated FastAPI security utilities
- Enhanced error handling
- Added token expiration management

### Models
- Updated to use SQLAlchemy Mapped columns
- Added data integrity constraints
- Refined model relationships
- Improved code clarity

### Resources
Refactored all resources (Authentication, Plan, Recipe, Shopping List) to:
- Use FastAPI instead of Flask-RESTful
- Implement Pydantic models for validation
- Support async operations
- Enhance error handling

### Services
Updated all service layers (Recipe, Shopping List, User Auth, User Plan) to:
- Support async architecture
- Optimize data access patterns
- Implement non-blocking operations
- Improve code maintainability

### Token Storage
- Implemented abstract base class for operations
- Added Redis integration for token management
- Used async methods for storage/retrieval
- Added comprehensive error handling

## Testing
1. Update environment variables for PostgreSQL and Redis
2. Run `docker-compose up --build`
3. Verify API endpoints through Swagger UI at `/docs`
